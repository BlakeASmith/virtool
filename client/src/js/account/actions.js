/**
 *
 *
 * @copyright 2017 Government of Canada
 * @license MIT
 * @author igboyes
 *
 */
import {
    WS_UPDATE_ACCOUNT,
    GET_ACCOUNT,
    UPDATE_ACCOUNT_SETTINGS,
    CREATE_API_KEY,
    UPDATE_API_KEY,
    REMOVE_API_KEY,
    LOGOUT
} from "../actionTypes";


export function wsUpdateAccount (data) {
    return {
        type: WS_UPDATE_ACCOUNT,
        data
    };
}

export function getAccount () {
    return {
        type: GET_ACCOUNT.REQUESTED
    }
}

export function updateAccountSettings (update) {
    return {
        type: UPDATE_ACCOUNT_SETTINGS.REQUESTED,
        update
    };
}

export function createAPIKey (name, permissions, callback) {
    return {
        type: CREATE_API_KEY.REQUESTED,
        name,
        permissions,
        callback
    };
}

export function updateAPIKey (keyId, permissions) {
    return {
        type: UPDATE_API_KEY.REQUESTED,
        keyId,
        permissions
    };
}

export function removeAPIKey (keyId) {
    return {
        type: REMOVE_API_KEY.REQUESTED,
        keyId
    };
}

export function logout () {
    return {
        type: LOGOUT.REQUESTED
    }
}
