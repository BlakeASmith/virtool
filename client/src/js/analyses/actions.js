import {
    WS_INSERT_ANALYSIS,
    WS_UPDATE_ANALYSIS,
    WS_REMOVE_ANALYSIS,
    FIND_ANALYSES,
    GET_ANALYSIS,
    ANALYZE,
    BLAST_NUVS,
    REMOVE_ANALYSIS,
    CLEAR_ANALYSIS,
    COLLAPSE_ANALYSIS,
    SET_PATHOSCOPE_FILTER,
    TOGGLE_SORT_PATHOSCOPE_DESCENDING,
    TOGGLE_SHOW_PATHOSCOPE_READS,
    SET_ACTIVE_HIT_ID,
    TOGGLE_FILTER_ORFS,
    TOGGLE_FILTER_SEQUENCES,
    SET_SEARCH_IDS,
    SET_ANALYSIS_SORT_KEY,
    TOGGLE_RESULT_EXPANDED,
    CLEAR_ANALYSES
} from "../app/actionTypes";
import { simpleActionCreator } from "../utils/utils";

export const wsInsertAnalysis = data => ({
    type: WS_INSERT_ANALYSIS,
    data
});

/**
 * Returns an action that should be dispatched when a analysis document is updated via websocket.
 *
 * @func
 * @param data {object} update data passed in the websocket message
 * @returns {object}
 */
export const wsUpdateAnalysis = data => ({
    type: WS_UPDATE_ANALYSIS,
    data
});

/**
 * Returns an action that should be dispatched when a analysis document is removed via websocket.
 *
 * @func
 * @param data {string} the id for the specific analysis
 * @returns {object}
 */
export const wsRemoveAnalysis = data => ({
    type: WS_REMOVE_ANALYSIS,
    data
});

export const collapseAnalysis = simpleActionCreator(COLLAPSE_ANALYSIS);

export const setActiveHitId = id => ({
    type: SET_ACTIVE_HIT_ID,
    id
});

export const setSearchIds = ids => ({
    type: SET_SEARCH_IDS,
    ids
});

export const setAnalysisSortKey = sortKey => ({
    type: SET_ANALYSIS_SORT_KEY,
    sortKey
});

export const setPathoscopeFilter = key => ({
    type: SET_PATHOSCOPE_FILTER,
    key
});

export const toggleFilterORFs = simpleActionCreator(TOGGLE_FILTER_ORFS);

export const toggleFilterSequences = simpleActionCreator(TOGGLE_FILTER_SEQUENCES);

export const togglePathoscopeSortDescending = simpleActionCreator(TOGGLE_SORT_PATHOSCOPE_DESCENDING);

export const toggleResultExpanded = id => ({
    type: TOGGLE_RESULT_EXPANDED,
    id
});

export const toggleShowPathoscopeReads = simpleActionCreator(TOGGLE_SHOW_PATHOSCOPE_READS);

export const findAnalyses = (sampleId, term, page) => ({
    type: FIND_ANALYSES.REQUESTED,
    sampleId,
    term,
    page
});

/**
 * Returns action that can trigger an API call for retrieving a specific analysis.
 *
 * @func
 * @param analysisId {string} unique analysis id
 * @returns {object}
 */
export const getAnalysis = analysisId => ({
    type: GET_ANALYSIS.REQUESTED,
    analysisId
});

export const clearAnalyses = simpleActionCreator(CLEAR_ANALYSES);
export const clearAnalysis = simpleActionCreator(CLEAR_ANALYSIS);

/**
 * Returns action that can trigger an API call for sample analysis.
 *
 * @func
 * @param sampleId {string} unique sample id
 * @param refId {string} unique id for a reference
 * @param algorithm {string} algorithm type
 * @param userId {string} the id of the requesting user
 * @returns {object}
 */
export const analyze = (sampleId, refId, algorithm, subtractionId, userId) => ({
    type: ANALYZE.REQUESTED,
    algorithm,
    refId,
    sampleId,
    subtractionId,
    userId
});

/**
 * Returns action that can trigger an API call for BLASTing NuV analysis contigs.
 *
 * @func
 * @param analysisId {string} unique analysis id
 * @param sequenceIndex {number} index of the sequence
 * @returns {object}
 */
export const blastNuvs = (analysisId, sequenceIndex) => ({
    type: BLAST_NUVS.REQUESTED,
    analysisId,
    sequenceIndex
});

/**
 * Returns action that can trigger an API call for removing an analysis.
 *
 * @func
 * @param analysisId {string} unique analysis id
 * @returns {object}
 */
export const removeAnalysis = analysisId => ({
    type: REMOVE_ANALYSIS.REQUESTED,
    analysisId
});
