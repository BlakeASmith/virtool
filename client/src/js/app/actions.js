import { PUSH_STATE } from "./actionTypes";

export const pushState = state => ({
    type: PUSH_STATE,
    state
});
