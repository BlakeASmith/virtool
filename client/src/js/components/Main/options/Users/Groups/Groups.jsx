/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports Groups
 */

'use strict';

var _ = require('lodash');
var React = require('react');
var Row = require('react-bootstrap/lib/Row');
var Col = require('react-bootstrap/lib/Col');
var Modal = require('react-bootstrap/lib/Modal');
var Panel = require('react-bootstrap/lib/Panel');
var ListGroup = require('react-bootstrap/lib/ListGroup');

var Add = require('./Add.jsx');
var Permissions = require('./Permissions.jsx');
var Icon = require('virtool/js/components/Base/Icon.jsx');
var ListGroupItem = require('virtool/js/components/Base/PushListGroupItem.jsx');

/**
 * Renders either a table describing the sessions associated with the user or a panel with a message indicating no
 * sessions are associated with that user.
 *
 * @class
 */
var Groups = React.createClass({

    getInitialState: function () {
        return {
            documents: this.getEntries(),
            activeGroupIndex: 0
        };
    },

    select: function (newActiveGroupIndex) {
        this.setState({activeGroupIndex: newActiveGroupIndex});
    },

    getEntries: function () {
        return _.sortBy(dispatcher.collections.groups.documents, '_id');
    },

    componentDidMount: function () {
        dispatcher.collections.groups.on('change', this.update);
    },

    componentWillUnmount: function () {
        dispatcher.collections.groups.off('change', this.update);
    },

    update: function () {
        this.setState({documents: this.getEntries()});
    },

    remove: function (groupName) {
        dispatcher.collections.groups.request('remove_group', {
            _id: groupName
        });
    },

    render: function () {

        var groupItemComponents = this.state.documents.map(function (document, index) {
            var props = {
                key: document._id,
                className: this.state.activeGroupIndex == index ? 'band': null,
                onClick: function () {this.select(index)}.bind(this)
            };

            var callback = function () {
                this.remove(document._id);
            }.bind(this);

            var removeIcon;

            if (document._id !== 'limited' && document._id !== 'administrator') {
                removeIcon = <Icon name='remove' className='pull-right' onClick={callback} />;
            }

            return (
                <ListGroupItem {...props}>
                    {_.capitalize(document._id)}
                    {removeIcon}
                </ListGroupItem>
            );

        }, this);

        var activeGroup = this.state.documents[this.state.activeGroupIndex];

        return (
            <Modal dialogClassName='modal-md' show={this.props.show} onHide={this.props.onHide}>
                <Modal.Header>
                    User Groups
                </Modal.Header>
                <Modal.Body>
                    <Row>
                        <Col md={6}>
                            <Add collection={dispatcher.collections.groups} />
                            {groupItemComponents}
                        </Col>
                        <Col md={6}>
                            <Permissions
                                groupName={activeGroup._id}
                                permissions={activeGroup.permissions}
                                collection={dispatcher.collections.groups}
                            />
                        </Col>
                    </Row>
                </Modal.Body>
            </Modal>
        );
    }
});

module.exports = Groups;
