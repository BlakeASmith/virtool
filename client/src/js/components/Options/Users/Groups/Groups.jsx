/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports Groups
 */

'use strict';

var _ = require('lodash');
var React = require('react');
var Row = require('react-bootstrap/lib/Row');
var Col = require('react-bootstrap/lib/Col');
var Modal = require('react-bootstrap/lib/Modal');
var Panel = require('react-bootstrap/lib/Panel');
var ListGroup = require('react-bootstrap/lib/ListGroup');

var Utils = require('virtool/js/Utils');
var Add = require('./Add.jsx');
var Permissions = require('./Permissions.jsx');
var Icon = require('virtool/js/components/Base/Icon.jsx');
var ListGroupItem = require('virtool/js/components/Base/PushListGroupItem.jsx');

/**
 * Renders either a table describing the sessions associated with the user or a panel with a message indicating no
 * sessions are associated with that user.
 *
 * @class
 */
var Groups = React.createClass({

    getInitialState: function () {
        var documents = this.getEntries();

        return {
            documents: documents,
            activeId: documents[0]._id
        };
    },

    select: function (groupId) {
        this.setState({activeId: groupId});
    },

    getEntries: function () {
        return _.sortBy(dispatcher.db.groups.documents, '_id');
    },

    componentDidMount: function () {
        dispatcher.db.groups.on('change', this.update);
    },

    componentWillUnmount: function () {
        dispatcher.db.groups.off('change', this.update);
    },

    update: function () {
        var newDocuments = this.getEntries();

        this.setState({
            documents: newDocuments,
            activeId: Utils.getNewActiveId(this.state.activeId, this.state.documents, newDocuments)
        });
    },

    remove: function (groupName) {
        dispatcher.db.groups.request('remove_group', {
            _id: groupName
        });
    },

    render: function () {

        var groupItemComponents = this.state.documents.map(function (document) {
            var props = {
                key: document._id,
                className: this.state.activeId == document._id ? 'band': null,
                onClick: function () {this.select(document._id)}.bind(this)
            };

            var callback = function () {
                this.remove(document._id);
            }.bind(this);

            var removeIcon;

            if (document._id !== 'limited' && document._id !== 'administrator') {
                removeIcon = <Icon name='remove' className='pull-right' onClick={callback} />;
            }

            return (
                <ListGroupItem {...props}>
                    {_.capitalize(document._id)}
                    {removeIcon}
                </ListGroupItem>
            );

        }, this);

        var activeGroup = _.find(this.state.documents, {_id: this.state.activeId});

        return (
            <Modal dialogClassName='modal-md' show={this.props.show} onHide={this.props.onHide}>
                <Modal.Header>
                    User Groups
                </Modal.Header>
                <Modal.Body>
                    <Row>
                        <Col md={6}>
                            <Add collection={dispatcher.db.groups} />
                            {groupItemComponents}
                        </Col>
                        <Col md={6}>
                            <Permissions
                                groupName={activeGroup._id}
                                permissions={activeGroup.permissions}
                                collection={dispatcher.db.groups}
                            />
                        </Col>
                    </Row>
                </Modal.Body>
            </Modal>
        );
    }
});

module.exports = Groups;
