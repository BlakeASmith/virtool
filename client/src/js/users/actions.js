/**
 *
 *
 * @copyright 2017 Government of Canada
 * @license MIT
 * @author igboyes
 *
 */

import {
    LIST_USERS,
    FILTER_USERS,
    CREATE_USER,
    SET_PASSWORD,
    SET_PRIMARY_GROUP,
    SET_FORCE_RESET,
    ADD_USER_TO_GROUP,
    REMOVE_USER_FROM_GROUP
} from "../actionTypes";

export const listUsers = () => {
    return {
        type: LIST_USERS.REQUESTED
    };
};

export const filterUsers = (term) => {
    return {
        type: FILTER_USERS,
        term
    };
};

export const createUser = (userId, password, forceReset) => {
    return {
        type: CREATE_USER.REQUESTED,
        userId,
        password,
        forceReset
    };
};

export const setForceReset = (userId, enabled) => {
    return {
        type: SET_FORCE_RESET.REQUESTED,
        userId: userId,
        enabled: enabled
    };
};

export const setPassword = (userId, password) => {
    return {
        type: SET_PASSWORD.REQUESTED,
        userId,
        password
    };
};

export const setPrimaryGroup = (userId, primaryGroup) => {
    return {
        type: SET_PRIMARY_GROUP.REQUESTED,
        userId,
        primaryGroup
    };
};

export const addUserToGroup = (userId, groupId) => {
    return {
        type: ADD_USER_TO_GROUP.REQUESTED,
        userId,
        groupId
    };
};

export const removeUserFromGroup = (userId, groupId) => {
    return {
        type: REMOVE_USER_FROM_GROUP.REQUESTED,
        userId,
        groupId
    };
};
