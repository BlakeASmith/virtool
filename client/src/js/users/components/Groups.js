import React from "react";
import { includes, map } from "lodash-es";
import { connect } from "react-redux";
import { Row, Col, Panel } from "react-bootstrap";
import { ListGroupItem, Checkbox } from "../../base";

import { addUserToGroup, removeUserFromGroup } from "../actions";

class UserGroup extends React.Component {

    handleClick = () => {
        const { groupId, userId } = this.props;

        if (this.props.toggled) {
            return this.props.removeFromGroup(userId, groupId);
        }

        this.props.addToGroup(userId, groupId);
    };

    render () {
        return (
            <Col xs={12} md={4} key={this.props.groupId}>
                <ListGroupItem
                    className="text-capitalize"
                    onClick={this.handleClick}
                    disabled={this.props.disabled}
                >
                    {this.props.groupId}
                    <Checkbox checked={this.props.toggled} pullRight />
                </ListGroupItem>
            </Col>
        );
    }
}

const UserGroups = ({ accountUserId, addToGroup, allGroups, memberGroups, removeFromGroup, userId }) => {

    const groupComponents = map(allGroups, groupId =>
        <UserGroup
            key={groupId}
            groupId={groupId}
            accountUserId={accountUserId}
            disabled={groupId === "administrator" && userId === accountUserId}
            toggled={includes(memberGroups, groupId)}
            onAddToGroup={addToGroup}
            onRemoveFromGroup={removeFromGroup}
        />
    );

    return (
        <Panel>
            <Row>
                {groupComponents}
            </Row>
        </Panel>
    );
};

const mapStateToProps = state => ({
    accountUserId: state.account.id,
    allGroups: map(state.groups.list, "id")
});

const mapDispatchToProps = dispatch => ({

    addToGroup: (userId, groupId) => {
        dispatch(addUserToGroup(userId, groupId));
    },

    removeFromGroup: (userId, groupId) => {
        dispatch(removeUserFromGroup(userId, groupId));
    }

});

const Container = connect(mapStateToProps, mapDispatchToProps)(UserGroups);

export default Container;
