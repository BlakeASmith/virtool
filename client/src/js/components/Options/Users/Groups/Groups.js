/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports Groups
 */

import React from "react";
import FlipMove from "react-flip-move"
import { capitalize, find } from "lodash-es";
import { Row, Col } from "react-bootstrap";
import { Modal, Icon, ListGroupItem } from "virtool/js/components/Base";

import Add from "./Add";
import Permissions from "./Permissions";

const getState = () => {
    const documents = this.getEntries();

    return {
        documents: documents,
        activeId: documents[0]._id
    };
};

/**
 * Renders either a table describing the sessions associated with the user or a panel with a message indicating no
 * sessions are associated with that user.
 *
 * @class
 */
export default class Groups extends React.Component {

    constructor (props) {
        super(props);
        this.state = getState();
    }

    static propTypes = {
        show: React.PropTypes.bool,
        onHide: React.PropTypes.func
    };

    componentDidMount () {
        dispatcher.db.groups.on("change", this.update);
    }

    componentWillUnmount () {
        dispatcher.db.groups.off("change", this.update);
    }

    getEntries = () => dispatcher.db.groups.chain().find().simplesort("_id").data();

    select = (groupId) => {
        this.setState({activeId: groupId});
    };

    update = () => {
        this.setState(getState());
    };

    remove = (groupName) => {
        dispatcher.db.groups.request("remove_group", {
            _id: groupName
        });
    };

    render () {

        const groupItemComponents = this.state.documents.map((document) => {
            const props = {
                key: document._id,
                active: this.state.activeId == document._id,
                onClick: function () {this.select(document._id)}.bind(this)
            };

            let removeIcon;

            if (document._id !== "limited" && document._id !== "administrator") {
                removeIcon = <Icon name="remove" className="pull-right" onClick={() => this.remove(document._id)} />;
            }

            return (
                <ListGroupItem {...props}>
                    {capitalize(document._id)}
                    {removeIcon}
                </ListGroupItem>
            );

        });

        const activeGroup = find(this.state.documents, {_id: this.state.activeId});

        return (
            <Modal dialogClassName="modal-md" show={this.props.show} onHide={this.props.onHide}>
                <Modal.Header>
                    User Groups
                </Modal.Header>
                <Modal.Body>
                    <Row>
                        <Col md={6}>
                            <FlipMove>
                                <Add collection={dispatcher.db.groups} />
                                {groupItemComponents}
                            </FlipMove>
                        </Col>
                        <Col md={6}>
                            <FlipMove leaveAnimation={false} duration={200}>
                                <Permissions
                                    key={activeGroup._id}
                                    groupName={activeGroup._id}
                                    permissions={activeGroup.permissions}
                                    collection={dispatcher.db.groups}
                                />
                            </FlipMove>
                        </Col>
                    </Row>
                </Modal.Body>
            </Modal>
        );
    }
}
