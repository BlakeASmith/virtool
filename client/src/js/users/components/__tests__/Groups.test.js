import { listGroups } from "../../../groups/actions";
import { editUser } from "../../actions";
import { UserGroups, mapStateToProps, mapDispatchToProps } from "../Groups";

describe("<UserGroups />", () => {
    let props;

    beforeEach(() => {
        props = {
            userId: "bob",
            allGroups: ["foo", "bar"],
            memberGroups: ["foo"],
            onList: jest.fn(),
            onEditGroup: jest.fn()
        };
    });

    it("renders", () => {
        const wrapper = shallow(<UserGroups {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders <NoneFound /> if no groups available", () => {
        props = {
            ...props,
            allGroups: [],
            memberGroups: []
        };
        const wrapper = shallow(<UserGroups {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("calls onList props when component mounts", () => {
        mount(<UserGroups {...props} />);
        expect(props.onList).toBeCalled();
    });

    describe("calls onEditGroup when group is edited", () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallow(<UserGroups {...props} />);
        });

        it("enables membership", () => {
            wrapper.instance().handleEdit("bar");
            expect(props.onEditGroup).toBeCalledWith("bob", ["foo", "bar"]);
        });

        it("disables membership", () => {
            wrapper.instance().handleEdit("foo");
            expect(props.onEditGroup).toBeCalledWith("bob", []);
        });
    });

    it("mapStateToProps return props", () => {
        const documents = [{ id: "foo" }];
        const detail = {
            id: "bob",
            groups: ["foo", "bar"]
        };
        const props = mapStateToProps({
            groups: {
                documents
            },
            users: {
                detail
            }
        });
        expect(props).toEqual({
            documents,
            memberGroups: detail.groups,
            userId: detail.id
        });
    });

    describe("mapDispatchToProps", () => {
        let dispatch;
        let props;

        beforeEach(() => {
            dispatch = jest.fn();
            props = mapDispatchToProps(dispatch);
        });

        it("returns onList", () => {
            props.onList();
            expect(dispatch).toBeCalledWith(listGroups());
        });

        it("returns onEditGroup", () => {
            const groups = ["foo", "bar"];
            props.onEditGroup("bob", groups);
            expect(dispatch).toBeCalledWith(editUser("bob", { groups }));
        });
    });
});
