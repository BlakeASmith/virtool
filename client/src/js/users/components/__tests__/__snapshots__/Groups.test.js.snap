// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserGroups /> renders <NoneFound /> if no groups available 1`] = `
<NoneFound
  noun="groups"
/>
`;

exports[`<UserGroups /> renders 1`] = `
<NoneFound
  noun="groups"
/>
`;
